### ISSUE TITLE
Object files shouldn't include a NAME section

### ISSUE THREAD

**Issue by sbc100** [WebAssembly/tool-conventions#37]
The NAME section of the object files currently duplicates information that is already stored in the import/export/linking section which is bad for several reasons:

1. Its redundant and wastes space
2. The information can get out of sync (which is the canonical name for a function?)

We already have a change out to remove the NAME section from the "symbol table": 
https://reviews.llvm.org/D42075

But I propose we take it one step further and simply don't emit a NAME section when writing the object file.

**Comment by sbc100** [comment]
@sunfishcode @NWilson

**Comment by NWilson** [comment]
Good idea. Could we do this in a separate commit, rather than stuff it into https://reviews.llvm.org/D42075?
 D42075 covers the internal representation of the NAME section, and I was hoping it would be merged _quicker_ if I avoided changing other things at once - but now it's delayed because I wasn't brave enough to just get rid of it!

(We can't change the format of the NAME section, since Chrome/Firefox parse it for use in their Wasm disassembler. LLD therefore has to emit it, therefore WasmObjectFile has to continue to parse it so that obj2yaml can be used to test LLD's output. So this discussion is purely about whether Clang's WasmObjectWriter produces a NAME section in the intermediate object files.)

1. True it's redundant, but it's not a big amount of data. Once we've got file names and line numbers and variable names included in the debug data, it's going to be a smaller percentage saving.
2. The canonical name for a function is the one in the NAME section (since a function body can be exported multiple times under different names). It shouldn't get out of sync - in fact this is in favour of emitting the NAME section. If we didn't emit a NAME section in Clang, LLD would "guess" the function's name from the first Symbol referencing it.

I'd be very happy to remove it, I agree it's not necessary. The first export (or symtab entry) referencing a function is a reliable name, for LLD to use when writing out the linked output.
